File: chat/tests/muc-call-presence.test.ts
import { afterEach, describe, expect, test } from "bun:test";
import {
  $mucCallParticipants,
  applyMucCallPresence,
  awaitNoOtherPreparing,
  awaitPreparingEcho,
  cancelMucCallPreparationWaiters,
  clearMucCallParticipant,
  clearMucCallParticipants,
  mucCallParticipantCounts,
  mucCallParticipantCount,
  pendingMucCallPreparationWaiterCountForTests,
} from "../src/lib/calls/muc-call-presence";

afterEach(() => {
  clearMucCallParticipants();
});

// XEP-0272 Muji presence — the namespace + element shape that the
// chat-side store consumes. The previous custom `urn:waddle:muc-call:0`
// extension has been retired; these tests exercise the active/preparing
// boolean pair that the WASM parser surfaces in `WasmMujiPresence`.
const activeMuji = { preparing: false, active: true } as const;
const preparingMuji = { preparing: true, active: false } as const;
const activePreparingMuji = { preparing: true, active: true } as const;

describe("applyMucCallPresence", () => {
  test("available presence with active Muji registers the nick", () => {
    applyMucCallPresence({
      from: "Room@MUC.Test/alice",
      presence_type: "available",
      muji: activeMuji,
    });
    expect($mucCallParticipants.get()).toEqual({
      "room@muc.test": ["alice"],
    });
    expect(mucCallParticipantCount("room@muc.test")).toBe(1);
    expect(mucCallParticipantCount("ROOM@MUC.TEST/resource")).toBe(1);
  });

  test("multiple occupants accumulate per room", () => {
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muji: activeMuji,
    });
    applyMucCallPresence({
      from: "room@muc.test/bob",
      presence_type: "available",
      muji: activeMuji,
    });
    expect($mucCallParticipants.get()["room@muc.test"]).toEqual([
      "alice",
      "bob",
    ]);
  });

  test("preparing-only Muji does NOT count as active (XEP-0272 §Joining two-phase flow)", () => {
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muji: preparingMuji,
    });
    expect($mucCallParticipants.get()["room@muc.test"]).toBeUndefined();
  });

  test("preparing-only Muji preserves an existing active nick for same-nick sibling joins", () => {
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muji: activeMuji,
    });
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muji: preparingMuji,
    });
    expect($mucCallParticipants.get()["room@muc.test"]).toEqual(["alice"]);
  });

  test("plain same-nick sibling clear preserves the exact active owner", () => {
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/desktop",
      muji: activeMuji,
    });
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/mobile",
    });
    expect($mucCallParticipants.get()["room@muc.test"]).toEqual(["alice"]);

    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/desktop",
    });
    expect($mucCallParticipants.get()["room@muc.test"]).toBeUndefined();
  });

  test("clearing one active same-nick owner preserves another active owner", () => {
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/desktop",
      muji: activeMuji,
    });
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/mobile",
      muji: activeMuji,
    });

    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/desktop",
    });
    expect($mucCallParticipants.get()["room@muc.test"]).toEqual(["alice"]);

    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/mobile",
    });
    expect($mucCallParticipants.get()["room@muc.test"]).toBeUndefined();
  });

  test("local exact clear preserves active same-nick siblings", () => {
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/desktop",
      muji: activeMuji,
    });
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/mobile",
      muji: activeMuji,
    });

    clearMucCallParticipant("room@muc.test", "alice", "alice@waddle.test/desktop");
    expect($mucCallParticipants.get()["room@muc.test"]).toEqual(["alice"]);

    clearMucCallParticipant("room@muc.test", "alice", "alice@waddle.test/mobile");
    expect($mucCallParticipants.get()["room@muc.test"]).toBeUndefined();
  });

  test("departed active same-nick resource clears before preparing sibling replay", () => {
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/desktop",
      muji: activeMuji,
    });
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/mobile",
      muji: preparingMuji,
    });
    expect($mucCallParticipants.get()["room@muc.test"]).toEqual(["alice"]);

    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/desktop",
    });
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/mobile",
      muji: preparingMuji,
    });

    expect($mucCallParticipants.get()["room@muc.test"]).toBeUndefined();
  });

  test("active+preparing Muji counts as active while preserving the setup signal", () => {
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muji: activePreparingMuji,
    });
    expect($mucCallParticipants.get()["room@muc.test"]).toEqual(["alice"]);
  });

  test("available presence WITHOUT the Muji extension removes a previously-registered nick (XEP-0272 §Leaving)", () => {
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muji: activeMuji,
    });
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
    });
    expect($mucCallParticipants.get()["room@muc.test"]).toBeUndefined();
  });

  test("unavailable presence removes the nick (occupant left the room entirely)", () => {
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muji: activeMuji,
    });
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "unavailable",
    });
    expect($mucCallParticipants.get()["room@muc.test"]).toBeUndefined();
  });

  test("rooms are tracked independently", () => {
    applyMucCallPresence({
      from: "room-a@muc.test/alice",
      presence_type: "available",
      muji: activeMuji,
    });
    applyMucCallPresence({
      from: "room-b@muc.test/carol",
      presence_type: "available",
      muji: activeMuji,
    });
    expect($mucCallParticipants.get()).toEqual({
      "room-a@muc.test": ["alice"],
      "room-b@muc.test": ["carol"],
    });
  });

  test("re-adding an already-present nick is idempotent", () => {
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muji: activeMuji,
    });
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muji: activeMuji,
    });
    expect($mucCallParticipants.get()["room@muc.test"]).toEqual(["alice"]);
  });

  test("removing a never-added nick is a no-op", () => {
    applyMucCallPresence({
      from: "room@muc.test/ghost",
      presence_type: "unavailable",
    });
    expect($mucCallParticipants.get()).toEqual({});
  });

  test("ignores presences missing a `from` or nick", () => {
    applyMucCallPresence({
      presence_type: "available",
      muji: activeMuji,
    });
    applyMucCallPresence({
      from: "room@muc.test",
      presence_type: "available",
      muji: activeMuji,
    });
    expect($mucCallParticipants.get()).toEqual({});
  });

  test("clearMucCallParticipants drops every tracked room", () => {
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muji: activeMuji,
    });
    clearMucCallParticipants();
    expect($mucCallParticipants.get()).toEqual({});
  });

  test("clearMucCallParticipants rejects pending preparing echo waiters", async () => {
    const result = awaitPreparingEcho("room@muc.test", "alice", 5000)
      .then(() => "resolved" as const)
      .catch((err) => err);
    clearMucCallParticipants();
    const err = await result;
    expect(err).toBeInstanceOf(Error);
    expect(String(err.message)).toContain("cancelled");
  });

  test("mucCallParticipantCounts normalizes room keys for sidebar/header indicators", () => {
    expect(mucCallParticipantCounts({
      "Room@MUC.Test/web": ["alice"],
      "room@muc.test": ["bob"],
      "other@muc.test": ["carol"],
    })).toEqual({
      "room@muc.test": 2,
      "other@muc.test": 1,
    });
  });

  test("registers the local user's own nick when their sibling resource starts a call (regression: cross-instance indicator)", () => {
    // Scenario: alice is signed in on two browser instances (web +
    // mobile). Both have joined `room@muc.test` with the same nick
    // "alice" — XEP-0045 §7.2 same-bare multi-session join. Web
    // starts a call; the server reflects the Muji presence to BOTH
    // sessions per XEP-0045 §7.1. Mobile's wasm bridge surfaces the
    // reflection here.
    //
    // The store MUST register "alice" as an in-call nick on mobile
    // too — otherwise mobile's sidebar chip never lights up and the
    // user has to manually re-discover the call. This was the
    // user-visible symptom of the server-side delivery bug
    // (`muc_update.rs` routing self-bare recipients onto the
    // sender's WebSocket via `responses` instead of the sibling's
    // via the connection registry).
    //
    // Pin the contract: a presence whose `from` reflects our own
    // identity is NOT filtered out. The store has no concept of
    // "self" — and intentionally so. A future refactor that adds an
    // implicit self-filter would silently re-break the multi-
    // instance indicator without any other test catching it.
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muji: activeMuji,
    });
    expect($mucCallParticipants.get()["room@muc.test"]).toEqual(["alice"]);
    expect(mucCallParticipantCount("room@muc.test")).toBe(1);

    // The matching XEP-0272 §Leaving reflection (empty `<muji/>`
    // gets stripped by the server, so the chat side sees an
    // available presence with no muji field) must clear the entry
    // even though it reflects our own identity.
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
    });
    expect($mucCallParticipants.get()["room@muc.test"]).toBeUndefined();
  });
});

describe("awaitPreparingEcho (XEP-0272 §Joining MUST)", () => {
  test("resolves when a matching preparing-only presence echo arrives", async () => {
    // The pre-call flow registers an echo waiter then fires the
    // preparing presence; the MUC echoes it back; the waiter
    // resolves so beginMucCall can proceed.
    const wait = awaitPreparingEcho("room@muc.test", "alice", 5000);
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muji: preparingMuji,
    });
    await wait;
    // No assertion needed — the await itself is the test. If the
    // echo never fires the listener, the test times out.
  });

  test("does NOT resolve on a different nick's preparing echo", async () => {
    // Two participants prepare at the same time; alice's waiter
    // must NOT resolve on bob's echo.
    const aliceWait = awaitPreparingEcho("room@muc.test", "alice", 200);
    let aliceResolved = false;
    aliceWait.then(() => {
      aliceResolved = true;
    });
    applyMucCallPresence({
      from: "room@muc.test/bob",
      presence_type: "available",
      muji: preparingMuji,
    });
    // Give the microtask queue a chance to run before checking.
    await new Promise((r) => setTimeout(r, 10));
    expect(aliceResolved).toBe(false);
    // Now alice's echo arrives:
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muji: preparingMuji,
    });
    await aliceWait;
    expect(aliceResolved).toBe(true);
  });

  test("does NOT resolve on a same-nick sibling resource's preparing echo", async () => {
    const aliceWait = awaitPreparingEcho(
      "room@muc.test",
      "alice",
      200,
      "alice@waddle.test/web",
    );
    let aliceResolved = false;
    void aliceWait.then(() => {
      aliceResolved = true;
    }).catch(() => undefined);

    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/mobile",
      muji: preparingMuji,
    });
    await new Promise((r) => setTimeout(r, 10));
    expect(aliceResolved).toBe(false);

    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/web",
      muji: preparingMuji,
    });
    await aliceWait;
    expect(aliceResolved).toBe(true);
  });

  test("does NOT resolve on a content-only presence", async () => {
    // A content-only presence is NOT a preparing echo; the waiter
    // must continue to block until the MUC echoes preparing.
    const wait = awaitPreparingEcho("room@muc.test", "alice", 200);
    let resolved = false;
    void wait.then(() => {
      resolved = true;
    }).catch(() => undefined);
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muji: activeMuji, // active, not preparing
    });
    await new Promise((r) => setTimeout(r, 10));
    expect(resolved).toBe(false);
    await expect(wait).rejects.toThrow("Timed out waiting for Muji preparing presence echo");
  });

  test("resolves on active+preparing presence because preparing remains the setup signal", async () => {
    const wait = awaitPreparingEcho(
      "room@muc.test",
      "alice",
      200,
      "alice@waddle.test/web",
    );
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/web",
      muji: activePreparingMuji,
    });
    await wait;
  });

  test("rejects on timeout when no echo arrives", async () => {
    // Missing preparing echo is fatal for strict XEP-0272 setup:
    // beginMucCall must roll back instead of proceeding with room
    // state the MUC never reflected.
    await expect(awaitPreparingEcho("room@muc.test/nobody", "alice", 50))
      .rejects.toThrow("Timed out waiting for Muji preparing presence echo");
  });
});

describe("awaitNoOtherPreparing", () => {
  test("same-nick sibling resources still count as other preparing participants", async () => {
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/mobile",
      muji: preparingMuji,
    });

    const wait = awaitNoOtherPreparing(
      "room@muc.test",
      "alice",
      200,
      "alice@waddle.test/web",
    );
    let resolved = false;
    void wait.then(() => {
      resolved = true;
    });
    await new Promise((r) => setTimeout(r, 10));
    expect(resolved).toBe(false);

    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/mobile",
      muji: activeMuji,
    });

    await wait;
    expect(resolved).toBe(true);
  });

  test("active+preparing same-nick sibling remains a blocker until preparing clears", async () => {
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/mobile",
      muji: activePreparingMuji,
    });

    const wait = awaitNoOtherPreparing(
      "room@muc.test",
      "alice",
      200,
      "alice@waddle.test/web",
    );
    let resolved = false;
    void wait.then(() => {
      resolved = true;
    });
    await new Promise((r) => setTimeout(r, 10));
    expect(resolved).toBe(false);
    expect($mucCallParticipants.get()["room@muc.test"]).toEqual(["alice"]);

    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/mobile",
      muji: activeMuji,
    });

    await wait;
    expect(resolved).toBe(true);
  });

  test("plain same-nick sibling presence does not clear active+preparing owner", async () => {
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/mobile",
      muji: activePreparingMuji,
    });

    const wait = awaitNoOtherPreparing(
      "room@muc.test",
      "carol",
      200,
      "carol@waddle.test/web",
    );
    let resolved = false;
    void wait.then(() => {
      resolved = true;
    });
    await new Promise((r) => setTimeout(r, 10));
    expect(resolved).toBe(false);

    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/desktop",
    });
    await new Promise((r) => setTimeout(r, 10));
    expect(resolved).toBe(false);

    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/mobile",
      muji: activeMuji,
    });

    await wait;
    expect(resolved).toBe(true);
  });

  test("active+preparing exact owner leaves sibling preparing blockers intact", async () => {
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/mobile",
      muji: preparingMuji,
    });
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/tablet",
      muji: preparingMuji,
    });
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/mobile",
      muji: activePreparingMuji,
    });

    const wait = awaitNoOtherPreparing(
      "room@muc.test",
      "carol",
      200,
      "carol@waddle.test/web",
    );
    let resolved = false;
    void wait.then(() => {
      resolved = true;
    });
    await new Promise((r) => setTimeout(r, 10));
    expect(resolved).toBe(false);

    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/tablet",
      muji: activeMuji,
    });
    await new Promise((r) => setTimeout(r, 10));
    expect(resolved).toBe(false);

    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muc_jid: "alice@waddle.test/mobile",
      muji: activeMuji,
    });

    await wait;
    expect(resolved).toBe(true);
  });
});

describe("preparation waiter lifecycle (no dangling listeners/timers)", () => {
  test("a resolved echo waiter leaves no live listener", async () => {
    expect(pendingMucCallPreparationWaiterCountForTests()).toBe(0);
    const wait = awaitPreparingEcho("room@muc.test", "alice", 5000);
    expect(pendingMucCallPreparationWaiterCountForTests()).toBe(1);
    applyMucCallPresence({
      from: "room@muc.test/alice",
      presence_type: "available",
      muji: preparingMuji,
    });
    await wait;
    expect(pendingMucCallPreparationWaiterCountForTests()).toBe(0);
  });

  test("a timed-out echo waiter leaves no live listener", async () => {
    const wait = awaitPreparingEcho("room@muc.test", "alice", 20);
    expect(pendingMucCallPreparationWaiterCountForTests()).toBe(1);
    await expect(wait).rejects.toThrow("Timed out");
    expect(pendingMucCallPreparationWaiterCountForTests()).toBe(0);
  });

  test("cancel() on an abandoned echo waiter rejects and releases it immediately", async () => {
    const wait = awaitPreparingEcho("room@muc.test", "alice", 60_000);
    expect(pendingMucCallPreparationWaiterCountForTests()).toBe(1);
    const settled = wait.then(() => "resolved" as const).catch((err: Error) => err);
    wait.cancel();
    const result = await settled;
    expect(result).toBeInstanceOf(Error);
    // The map entry AND its 60s timer are gone the instant we cancel —
    // not parked until the (now-cancelled) timeout would have fired.
    expect(pendingMucCallPreparationWaiterCountForTests()).toBe(0);
  });

  test("cancel() on an abandoned no-other-preparing waiter rejects and releases it immediately", async () => {
    // A real OTHER preparing participant so the waiter actually parks.
    applyMucCallPresence({
      from: "room@muc.test/bob",
      presence_type: "available",
      muc_jid: "bob@waddle.test/web",
      muji: preparingMuji,
    });
    const wait = awaitNoOtherPreparing(
      "room@muc.test",
      "alice",
      60_000,
      "alice@waddle.test/web",
    );
    expect(pendingMucCallPreparationWaiterCountForTests()).toBe(1);
    const settled = wait.then(() => "resolved" as const).catch((err: Error) => err);
    wait.cancel(new Error("attempt abandoned"));
    const result = await settled;
    expect(result).toBeInstanceOf(Error);
    expect((result as Error).message).toContain("attempt abandoned");
    expect(pendingMucCallPreparationWaiterCountForTests()).toBe(0);
  });

  test("cancel() is a harmless no-op for an already-resolved no-other-preparing fast path", async () => {
    // No other preparing participants → resolves synchronously with no
    // registered listener. cancel() must not throw or leave state.
    const wait = awaitNoOtherPreparing("room@muc.test", "alice", 5000);
    await wait;
    expect(() => wait.cancel()).not.toThrow();
    expect(pendingMucCallPreparationWaiterCountForTests()).toBe(0);
  });

  test("cancelMucCallPreparationWaiters releases both waiter kinds for the attempt", async () => {
    applyMucCallPresence({
      from: "room@muc.test/bob",
      presence_type: "available",
      muc_jid: "bob@waddle.test/web",
      muji: preparingMuji,
    });
    const echo = awaitPreparingEcho("room@muc.test", "alice", 60_000, "alice@waddle.test/web");
    const noOther = awaitNoOtherPreparing("room@muc.test", "alice", 60_000, "alice@waddle.test/web");
    const echoSettled = echo.then(() => "resolved" as const).catch((err: Error) => err);
    const noOtherSettled = noOther.then(() => "resolved" as const).catch((err: Error) => err);
    expect(pendingMucCallPreparationWaiterCountForTests()).toBe(2);

    cancelMucCallPreparationWaiters("room@muc.test", "alice", new Error("call setup cancelled"));

    expect(await echoSettled).toBeInstanceOf(Error);
    expect(await noOtherSettled).toBeInstanceOf(Error);
    expect(pendingMucCallPreparationWaiterCountForTests()).toBe(0);
  });

  test("rapid re-registration of the same attempt does not accumulate listeners", async () => {
    const first = awaitPreparingEcho("room@muc.test", "alice", 60_000, "alice@waddle.test/web");
    const firstSettled = first.then(() => "resolved" as const).catch((err: Error) => err);
    const second = awaitPreparingEcho("room@muc.test", "alice", 60_000, "alice@waddle.test/web");
    const secondSettled = second.then(() => "resolved" as const).catch((err: Error) => err);
    // The second registration replaced the first — exactly one live waiter.
    expect(pendingMucCallPreparationWaiterCountForTests()).toBe(1);
    expect(await firstSettled).toBeInstanceOf(Error);

    second.cancel();
    expect(await secondSettled).toBeInstanceOf(Error);
    expect(pendingMucCallPreparationWaiterCountForTests()).toBe(0);
  });
});

File: apps/android/core/client/src/test/kotlin/social/waddle/android/client/calls/MucCallPresenceTest.kt
package social.waddle.android.client.calls

import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.async
import kotlinx.coroutines.test.advanceTimeBy
import kotlinx.coroutines.test.runCurrent
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * XEP-0272 Muji presence bookkeeping, mirroring the web
 * muc-call-presence.ts suite: §Joining/§Leaving add/remove rules,
 * preparing-only semantics, media aggregation, owner mapping, the
 * preparing-echo waiter, and the no-other-preparing wait.
 */
@OptIn(ExperimentalCoroutinesApi::class)
class MucCallPresenceTest {
    private val presence = MucCallPresence()

    // ── Membership rules ─────────────────────────────────────────────────────

    @Test
    fun `active muji presence adds the nick with owner and media`() {
        presence.applyMucCallPresence(mujiPresence("alice", active = true, mucJid = OWN_FULL))

        assertEquals(mapOf(ROOM_JID to setOf("alice")), presence.participants.value)
        assertEquals(mapOf(ROOM_JID to mapOf<String, String?>("alice" to OWN_FULL)), presence.owners.value)
        assertEquals(mapOf(ROOM_JID to MucCallMedia(audio = true, video = false)), presence.media.value)
    }

    @Test
    fun `presence without muji removes the nick per the leaving marker`() {
        presence.applyMucCallPresence(mujiPresence("alice", active = true))

        presence.applyMucCallPresence(mujiPresence("alice"))

        assertTrue(presence.participants.value.isEmpty())
        assertTrue(presence.owners.value.isEmpty())
        assertTrue(presence.media.value.isEmpty())
    }

    @Test
    fun `unavailable presence removes the nick`() {
        presence.applyMucCallPresence(mujiPresence("alice", active = true))

        presence.applyMucCallPresence(
            mujiPresence("alice", active = true, presenceType = "unavailable"),
        )

        assertTrue(presence.participants.value.isEmpty())
    }

    @Test
    fun `preparing-only muji never counts as call membership`() {
        presence.applyMucCallPresence(mujiPresence("alice", preparing = true))

        assertTrue(presence.participants.value.isEmpty())

        // Preparing does not CLEAR active membership from a sibling
        // resource either — it is purely the setup signal.
        presence.applyMucCallPresence(mujiPresence("alice", active = true, mucJid = OWN_FULL))
        presence.applyMucCallPresence(mujiPresence("alice", preparing = true, mucJid = PEER_FULL))
        assertEquals(mapOf(ROOM_JID to setOf("alice")), presence.participants.value)
    }

    @Test
    fun `duplicate and replayed presences are idempotent`() {
        presence.applyMucCallPresence(mujiPresence("alice", active = true))
        presence.applyMucCallPresence(mujiPresence("alice", active = true))
        assertEquals(mapOf(ROOM_JID to setOf("alice")), presence.participants.value)

        presence.applyMucCallPresence(mujiPresence("ghost"))
        assertEquals(mapOf(ROOM_JID to setOf("alice")), presence.participants.value)
    }

    @Test
    fun `media aggregates across active participants`() {
        presence.applyMucCallPresence(mujiPresence("alice", active = true, mucJid = OWN_FULL))
        presence.applyMucCallPresence(
            mujiPresence("bob", active = true, hasVideo = true, mucJid = PEER_FULL),
        )

        assertEquals(mapOf(ROOM_JID to MucCallMedia(audio = true, video = true)), presence.media.value)

        presence.applyMucCallPresence(mujiPresence("bob", mucJid = PEER_FULL))
        assertEquals(mapOf(ROOM_JID to MucCallMedia(audio = true, video = false)), presence.media.value)
    }

    @Test
    fun `removal is keyed by real jid when the room exposes it`() {
        presence.applyMucCallPresence(mujiPresence("alice", active = true, mucJid = OWN_FULL))
        presence.applyMucCallPresence(mujiPresence("alice", active = true, mucJid = PEER_FULL))

        // Dropping one resource's advertisement keeps the sibling's.
        presence.applyMucCallPresence(mujiPresence("alice", mucJid = OWN_FULL))
        assertEquals(mapOf(ROOM_JID to setOf("alice")), presence.participants.value)

        presence.applyMucCallPresence(mujiPresence("alice", mucJid = PEER_FULL))
        assertTrue(presence.participants.value.isEmpty())
    }

    @Test
    fun `clearParticipant drops the local occupant without an inbound presence`() {
        presence.applyMucCallPresence(mujiPresence("alice", active = true, mucJid = OWN_FULL))

        presence.clearParticipant(ROOM_JID, "alice", OWN_FULL)

        assertTrue(presence.participants.value.isEmpty())
    }

    // ── In-call flags ────────────────────────────────────────────────────────

    @Test
    fun `raised hand and mute markers track the presence flags`() {
        presence.applyMucCallPresence(
            mujiPresence("alice", active = true, handRaised = true, muted = true),
        )
        assertEquals(mapOf(ROOM_JID to setOf("alice")), presence.raisedHands.value)
        assertEquals(mapOf(ROOM_JID to setOf("alice")), presence.mutedNicks.value)

        presence.applyMucCallPresence(mujiPresence("alice", active = true, handRaised = false, muted = true))
        assertTrue(presence.raisedHands.value.isEmpty())
        assertEquals(mapOf(ROOM_JID to setOf("alice")), presence.mutedNicks.value)

        // Leaving the call clears all in-call state.
        presence.applyMucCallPresence(mujiPresence("alice"))
        assertTrue(presence.mutedNicks.value.isEmpty())
    }

    // ── Preparing-echo waiter ────────────────────────────────────────────────

    @Test
    fun `preparing echo resolves a waiter registered before the send`() = runTest {
        val waiter = presence.registerPreparingEchoWaiter(ROOM_JID, "alice", OWN_FULL)
        val wait = async { presence.awaitPreparingEcho(waiter, 2_000) }
        runCurrent()

        presence.applyMucCallPresence(mujiPresence("alice", preparing = true, mucJid = OWN_FULL))

        assertTrue(wait.await())
    }

    @Test
    fun `preparing echo waiter times out without an echo`() = runTest {
        val waiter = presence.registerPreparingEchoWaiter(ROOM_JID, "alice", OWN_FULL)
        val wait = async { presence.awaitPreparingEcho(waiter, 2_000) }
        runCurrent()

        advanceTimeBy(2_001)

        assertFalse(wait.await())
    }

    @Test
    fun `a sibling resource's echo does not satisfy our identity-keyed waiter`() = runTest {
        val waiter = presence.registerPreparingEchoWaiter(ROOM_JID, "alice", OWN_FULL)
        val wait = async { presence.awaitPreparingEcho(waiter, 2_000) }
        runCurrent()

        presence.applyMucCallPresence(mujiPresence("alice", preparing = true, mucJid = PEER_FULL))
        runCurrent()
        assertFalse(wait.isCompleted)

        advanceTimeBy(2_001)
        assertFalse(wait.await())
    }

    @Test
    fun `cancelPreparationWaiters fails a pending echo wait`() = runTest {
        val waiter = presence.registerPreparingEchoWaiter(ROOM_JID, "alice", OWN_FULL)
        val wait = async { presence.awaitPreparingEcho(waiter, 2_000) }
        runCurrent()

        presence.cancelPreparationWaiters(ROOM_JID, "alice")

        assertFalse(wait.await())
    }

    // ── No-other-preparing wait ──────────────────────────────────────────────

    @Test
    fun `awaitNoOtherPreparing resolves immediately when nobody is preparing`() = runTest {
        assertTrue(presence.awaitNoOtherPreparing(ROOM_JID, "alice", OWN_FULL, 2_000))
    }

    @Test
    fun `awaitNoOtherPreparing waits for the other occupant to finish preparation`() = runTest {
        presence.applyMucCallPresence(mujiPresence("bob", preparing = true, mucJid = PEER_FULL))
        val wait = async { presence.awaitNoOtherPreparing(ROOM_JID, "alice", OWN_FULL, 2_000) }
        runCurrent()
        assertFalse(wait.isCompleted)

        // Bob finishes preparing (contents declared, preparing dropped).
        presence.applyMucCallPresence(mujiPresence("bob", active = true, mucJid = PEER_FULL))

        assertTrue(wait.await())
    }

    @Test
    fun `awaitNoOtherPreparing times out while the other keeps preparing`() = runTest {
        presence.applyMucCallPresence(mujiPresence("bob", preparing = true, mucJid = PEER_FULL))
        val wait = async { presence.awaitNoOtherPreparing(ROOM_JID, "alice", OWN_FULL, 2_000) }
        runCurrent()

        advanceTimeBy(2_001)

        assertFalse(wait.await())
    }

    @Test
    fun `our own preparing entry does not block the no-other wait`() = runTest {
        presence.applyMucCallPresence(mujiPresence("alice", preparing = true, mucJid = OWN_FULL))

        assertTrue(presence.awaitNoOtherPreparing(ROOM_JID, "alice", OWN_FULL, 2_000))
    }
}
